python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py

```diff
@@ -0,0 +1,277 @@
+    reflect any subsequent writes.
+    """
+
+    effective_dt: datetime
+    last_event_id: Optional[int]
+
+
+class _AssetSliceCompatibleSubset(ValidAssetSubset): ...
+
+
+PartitionKey: TypeAlias = Optional[str]
+AssetPartition: TypeAlias = AssetKeyPartitionKey
+
+
+def _slice_from_subset(asset_graph_view: "AssetGraphView", subset: AssetSubset) -> "AssetSlice":
+    valid_subset = subset.as_valid(
+        asset_graph_view.asset_graph.get_partitions_def(subset.asset_key)
+    )
+    return AssetSlice(asset_graph_view, _AssetSliceCompatibleSubset(*valid_subset))
+
+
+class AssetSlice:
+    """An asset slice represents a set of partitions for a given asset key. It is
+    tied to a particular instance of an AssetGraphView, and is read-only.
+
+    With an asset slice you are able to traverse the set of partitions resident
+    in an asset graph at any point in time.
+
+    ```python
+        asset_graph_view_t0 = AssetGraphView.for_test(defs, effective_dt=some_date())
+
+        some_asset_slice = asset_graph_view_to.get_asset_slice(some_asset.key)
+
+        for parent_slice in some_asset_slice.get_parent_slices().values():
+            # do something with the parent slice
+    ```
+
+    AssetSlice is read-only and tied to a specific AssetGraphView. Therefore
+    we can aggressively cached methods and properties. However different methods
+    have different performance characterics so we have the following conventions:
+
+    Naming conventions
+    * Properties guaranteed to be fast.
+    * Methods prefixed with `get_` do some work in-memory but not hugely expensive
+    * Methods prefixed with `compute_` do potentially expensive work, like compute
+    * partition mappings and query the instance.
+    * Methods using "materialize" indicate that they fully materialize partition sets
+      These can potentially be very expensive if the underlying partition set has
+      an in-memory representation that involves large time windows. I.e. if
+      the underlying PartitionsSubset in the ValidAssetSubset is a TimeWindowPartitionsSubset
+      Usage of these methods should be avoided if possible if you are potentially
+      dealing with slices with large time-based partition windows.
+    """
+
+    def __init__(
+        self, asset_graph_view: "AssetGraphView", compatible_subset: _AssetSliceCompatibleSubset
+    ):
+        self._asset_graph_view = asset_graph_view
+        self._compatible_subset = compatible_subset
+
+    def convert_to_valid_asset_subset(self) -> ValidAssetSubset:
+        return self._compatible_subset
+
+    def materialize_partition_keys(self) -> AbstractSet[PartitionKey]:
+        return {ap.partition_key for ap in self._compatible_subset.asset_partitions}
+
+    @property
+    def asset_key(self) -> AssetKey:
+        return self._compatible_subset.asset_key
+
+    @cached_property
+    def parent_keys(self) -> AbstractSet[AssetKey]:
```

Comment:
Any particular reason this needs to be cached? It's not clear to me that fetching this value from the dictionary created by `cached_property` would be noticeably faster than fetching this value from the dictionary in our AssetGraph, and we'd be soaking up memory to do this.